File: rating-service/src/main/java/com/software/modsen/ratingservice/exceptionhandler/ExceptionHandlerControllerAdvice.java
package com.software.modsen.ratingservice.exceptionhandler;

import com.software.modsen.ratingservice.exception.DriverAlreadyHasRatingException;
import com.software.modsen.ratingservice.exception.FeignClientException;
import com.software.modsen.ratingservice.exception.NotFoundException;
import com.software.modsen.ratingservice.exception.RatingDriverException;
import org.springdoc.api.ErrorMessage;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.ResponseStatus;

import java.util.List;
import java.util.stream.Collectors;

@ControllerAdvice
public class ExceptionHandlerControllerAdvice {
    @ExceptionHandler({DriverAlreadyHasRatingException.class,
            RatingDriverException.class,
            FeignClientException.class})
    public ResponseEntity<ErrorMessage> handleErrors(RuntimeException exception) {
        return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(new ErrorMessage(exception.getMessage()));
    }

    @ExceptionHandler(NotFoundException.class)
    public ResponseEntity<ErrorMessage> handleNotFoundErrors(RuntimeException exception) {
        return ResponseEntity
                .status(HttpStatus.NOT_FOUND)
                .body(new ErrorMessage(exception.getMessage()));
    }

    @ResponseBody
    @ExceptionHandler(MethodArgumentNotValidException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public ValidationErrorResponse onMethodArgumentNotValidException(MethodArgumentNotValidException e) {
        final List<Violation> violations = e.getBindingResult().getFieldErrors().stream()
                .map(error -> new Violation(error.getField(), error.getDefaultMessage()))
                .collect(Collectors.toList());
        return new ValidationErrorResponse(violations);
    }
}

File: passenger-service/src/main/kotlin/com/software/modsen/passengerservice/exceptionhandler/ExceptionHandlerControllerAdvice.kt
package com.software.modsen.passengerservice.exceptionhandler

import com.software.modsen.passengerservice.exception.PassengerAlreadyExistsException
import com.software.modsen.passengerservice.exception.PassengerNotExistsException
import com.software.modsen.passengerservice.exception.PassengerRatingNotExistsException
import org.springdoc.api.ErrorMessage
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.ResponseStatus

@ControllerAdvice
class ExceptionHandlerControllerAdvice {
    @ExceptionHandler(
        PassengerNotExistsException::class,
        PassengerAlreadyExistsException::class,
        PassengerRatingNotExistsException::class
    )
    fun handleErrors(exception: RuntimeException): ResponseEntity<ErrorMessage> {
        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(ErrorMessage(exception.message))
    }

    @ResponseBody
    @ExceptionHandler(MethodArgumentNotValidException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun onMethodArgumentNotValidException(e: MethodArgumentNotValidException): ValidationErrorResponse {
        val violations = e.bindingResult.fieldErrors.map { error ->
            Violation(error.field, error.defaultMessage!!)
        }.toList()
        return ValidationErrorResponse(violations)
    }
}